provider/aws_sd.go

```diff
@@ -0,0 +1,626 @@
+		log.Info("All records are already up to date")
+		return nil
+	}
+
+	// convert updates to delete and create operation if applicable (updates not supported)
+	creates, deletes := p.updatesToCreates(changes)
+	changes.Delete = append(changes.Delete, deletes...)
+	changes.Create = append(changes.Create, creates...)
+
+	namespaces, err := p.ListNamespaces()
+	if err != nil {
+		return err
+	}
+
+	err = p.submitDeletes(namespaces, changes.Delete)
+	if err != nil {
+		return err
+	}
+
+	err = p.submitCreates(namespaces, changes.Create)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (p *AWSSDProvider) updatesToCreates(changes *plan.Changes) (creates []*endpoint.Endpoint, deletes []*endpoint.Endpoint) {
+	for i, old := range changes.UpdateOld {
+		current := changes.UpdateNew[i]
+
+		if !reflect.DeepEqual(old.Targets, current.Targets) || old.DNSName != current.DNSName {
+			// when targets or DNS name differ the old instances need to be de-registered first
+			deletes = append(deletes, old)
+		}
+
+		// always register (or re-register) instance with the current data
+		creates = append(creates, current)
+	}
+
+	return creates, deletes
+}
+
+func (p *AWSSDProvider) submitCreates(namespaces []*sd.NamespaceSummary, changes []*endpoint.Endpoint) error {
+	changesByNamespaceID := p.changesByNamespaceID(namespaces, changes)
+
+	for nsID, changeList := range changesByNamespaceID {
+		services, err := p.ListServicesByNamespaceID(aws.String(nsID))
+		if err != nil {
+			return err
+		}
+
+		for _, ch := range changeList {
+			_, srvName := p.parseHostname(ch.DNSName)
+
+			srv := services[srvName]
+			summary := serviceToServiceSummary(srv)
+			if srv == nil {
+				// when service is missing create a new one
+				summary, err = p.CreateService(&nsID, &srvName, ch)
+				if err != nil {
+					return err
+				}
+				// update local list of services
+				services, err = p.ListServicesByNamespaceID(aws.String(nsID))
+				if err != nil {
+					return err
+				}
+			} else {
+				// update service TTL when differs
+				if ch.RecordTTL.IsConfigured() && *srv.DnsConfig.DnsRecords[0].TTL != int64(ch.RecordTTL) {
+					err = p.UpdateService(srv, ch)
+					if err != nil {
+						return err
+					}
+				}
+			}
+
+			err = p.RegisterInstance(summary, ch)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+func (p *AWSSDProvider) submitDeletes(namespaces []*sd.NamespaceSummary, changes []*endpoint.Endpoint) error {
+	changesByNamespaceID := p.changesByNamespaceID(namespaces, changes)
+
+	for nsID, changeList := range changesByNamespaceID {
+		services, err := p.ListServicesByNamespaceID(aws.String(nsID))
+		if err != nil {
+			return err
+		}
+
+		for _, ch := range changeList {
+			hostname := ch.DNSName
+			_, srvName := p.parseHostname(hostname)
+
+			srv := serviceToServiceSummary(services[srvName])
+			if srv == nil {
+				return fmt.Errorf("FATAL ERROR: Service \"%s\" is missing", srvName)
```

Comment:
We should not be indicating whether the error is fatal here. Also errors returned should be lower cased - as a general convention. `return fmt.Errorf("service %s is missing", srvName)` is good enough

Reply:
Likewise in other places

Reply:
Will be fixed 👍